### PR TITLE
V6 routing updates

### PR DESCRIPTION
**Pull request by andreasohlund**
Updates the doco to include the latest v6 routing changes.

This also cleans up pub sub doco for all versions including a better explanation of automatic subscriptions

### CODE REVIEW

**Comment by andreasohlund** [comment]
This can't be pulled before https://github.com/Particular/NServiceBus/pull/2768 is in and uploaded to myget + refs updated for the v6 project

**Comment by andreasohlund** [comment]
@ramonsmits just realized that there is some overlaps btw this and https://github.com/Particular/docs.particular.net/pull/642 in regards to auto subscribe. Can you review this one and see what we can do?

**Comment by andreasohlund** [comment]
@SimonCropp this one should be good to go now, can you give it a once over?

**Comment by SimonCropp** [comment]
@andreasohlund can u please rebase

**Comment by andreasohlund** [comment]
done

File: Snippets/Snippets_6/packages.config [hunk 1]
```diff
@@ -1,7 +1,8 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="NServiceBus" version="6.0.0-unstable0601" targetFramework="net451" />
+  <package id="NServiceBus" version="6.0.0-unstable0612" targetFramework="net451" />
   <package id="NServiceBus.Callbacks" version="1.0.0-unstable0011" targetFramework="net451" />
   <package id="NServiceBus.Host" version="7.0.0-pullrequest0040" targetFramework="net451" />
+  <package id="NServiceBus.Host" version="6.0.0" targetFramework="net45" />
```
Comment (SimonCropp) [inline Snippets/Snippets_6/packages.config]: why does this ref two versions of the host?

File: nservicebus/messaging/publish-subscribe/how-to-pub-sub.md [hunk 2]
```diff
@@ -8,62 +8,47 @@ redirects:
...
-To manually subscribe and unsubscribe from a message:
+<!-- import InstancePublish -->
+
+If you are using interfaces to define your event contracts you need to set the message properties by passing in a lambda. NServiceBus will then generate a proxy and set those properties. 
+
+<!-- import InterfacePublish -->
+
+## Automatic subscriptions
+
+NServiceBus will automatically setup subscriptions for you. Messages matching the following criteria will be auto subscribed at startup by default.
```
Comment (SimonCropp) [inline nservicebus/messaging/publish-subscribe/how-to-pub-sub.md]: ` by default` is redundant

File: nservicebus/messaging/publish-subscribe/how-to-pub-sub.md [hunk 3]
```diff
@@ -8,62 +8,47 @@ redirects:
...
+
+1. Defined as a event either using `IEvent` or by the `.DefiningEventsAs` convention.
+2. At least one [message handler](/nservicebus/handlers/) for the given message
```
Comment (SimonCropp) [inline nservicebus/messaging/publish-subscribe/how-to-pub-sub.md]: i think this is missing `exists ` and should be 

> At least one [message handler](/nservicebus/handlers/) exists for the given message

File: nservicebus/messaging/publish-subscribe/how-to-pub-sub.md [hunk 4]
```diff
@@ -8,62 +8,47 @@ redirects:
...
+3. Has routing specified. Note that this only applies to transports that doesn't support Pub/Sub natively. Examples are Msmq, Sqlserver and Azure Storage Queues. See below for more details.
```
Comment (SimonCropp) [inline nservicebus/messaging/publish-subscribe/how-to-pub-sub.md]: > transports that dont

File: nservicebus/messaging/publish-subscribe/how-to-pub-sub.md [hunk 5]
```diff
@@ -8,62 +8,47 @@ redirects:
...
-    
-    <!-- To register a specific type in an assembly: -->
-    <add Assembly="assembly" Type="type fullname" Endpoint="queue@machinename" />
-  </MessageEndpointMappings>
-</UnicastBusConfig>
-```
+#### Exclude sagas from auto subscribe
 
-And a `UnicastBusConfig` config section.  
+Before V4 events that where only handled by sagas where not subscribed to by default. You can opt-in to the old behavior using:
```
Comment (SimonCropp) [inline nservicebus/messaging/publish-subscribe/how-to-pub-sub.md]: `Version 4` not `V4`

File: nservicebus/messaging/publish-subscribe/how-to-pub-sub.md [hunk 6]
```diff
@@ -8,62 +8,47 @@ redirects:
...
-  <configSections>
-    <section name="UnicastBusConfig" type="NServiceBus.Config.UnicastBusConfig, NServiceBus.Core" />
-```
+<!-- import DoNotAutoSubscribeSagas -->
 
-When subscribing to a message, you will probably have a [message handler](/nservicebus/handlers/) for it. If you do, and have the `UnicastBusConfig` section mentioned above, you do not have to write `Bus.Subscribe`, as NServiceBus invokes it automatically for you.
+#### Auto subscribe to plain messages
 
-## Disabling auto-subscription
+Before V4 all messages not defined as a command using `ICommand` or the `.DefiningCommandsAs` convention where automatically subscribed. You can opt-in to the old behavior using:
```
Comment (SimonCropp) [inline nservicebus/messaging/publish-subscribe/how-to-pub-sub.md]: `Version 4`

File: nservicebus/upgrades/5to6.md [hunk 7]
```diff
@@ -136,3 +136,7 @@ When you wanted to use the `IBuilder` interface outside the infrastructure of NS
 `var builder = ((UnicastBus)bus).Builder`
 
 This is no longer supported. We advise to, instead of using `IBuilder` directly, use dependency injection via your container of choice.
+
+## AutoSubscribe
+
+The configuration option `DoNotRequireExplicitRouting()` has been obsoleted since transports with support for centralized pubsub will default this to true. And transports with message driven pubsub like MSMQ, SqlServer and AzureStorageQueues will not subscribe properly if there is no routing specified. If you used this option you can safely be removed it.
```
Comment (SimonCropp) [inline nservicebus/upgrades/5to6.md]: > default this to true

u need to clarify why "true" is

Comment (SimonCropp) [inline nservicebus/upgrades/5to6.md]: Dont start a sentence with an `And`

Comment (SimonCropp) [inline nservicebus/upgrades/5to6.md]: put ` like MSMQ, SqlServer and AzureStorageQueues` in brackets

File: nservicebus/messaging/publish-subscribe/how-to-pub-sub.md (hunk 3 above)
Comment (SimonCropp) [inline nservicebus/messaging/publish-subscribe/how-to-pub-sub.md]: do we need to call out sagas here as well? or is that assumed using the word Handler?

File: Snippets/Snippets_6/Snippets_6.csproj [hunk 8]
```diff
@@ -138,4 +147,8 @@
     </ProjectReference>
   </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <PropertyGroup>
+    <StartAction>Program</StartAction>
+    <StartProgram>$(ProjectDir)$(OutputPath)NServiceBus.Host.exe</StartProgram>
```
Comment (SimonCropp) [inline Snippets/Snippets_6/Snippets_6.csproj]: please remove

File: Snippets/Snippets_6/Routing/EndpointMapping/ConfigurationSource/ConfigurationSource.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+                config.MessageEndpointMappings.Add(
+                    new MessageEndpointMapping
+                    {
+                        AssemblyName = "assembly",
+                        Endpoint = "queue@machinename"
+                    });
+                return config as T;
+            }
+
+            // To in app.config for other sections not defined in this method, otherwise return null.
```
Comment (SimonCropp) [inline Snippets/Snippets_6/Routing/EndpointMapping/ConfigurationSource/ConfigurationSource.cs]: `To in` ?

File: Snippets/Snippets_5/Routing/BasicOperations.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            bus.Publish(message);
+            #endregion
+
+        }
+
+        public void InterfaceMessage()
+        {
+            IBus bus = null;
+            #region InterfacePublish
+            var message = new MyEvent { SomeProperty = "Hello world" };
```
Comment (SimonCropp) [inline Snippets/Snippets_5/Routing/BasicOperations.cs]: why r we newing up a instance in a InterfacePublish snippet?

File: Snippets/Snippets_4/Routing/BasicOperations.cs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            bus.Publish(message);
+            #endregion
+
+        }
+
+        public void InterfaceMessage()
+        {
+            IBus bus = null;
+            #region InterfacePublish
+            var message = new MyEvent { SomeProperty = "Hello world" };
```
Comment (SimonCropp) [inline Snippets/Snippets_4/Routing/BasicOperations.cs]: newing up an interface?

File: Snippets/Snippets_3/Routing/BasicOperations.cs [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+            bus.Publish(message);
+            #endregion
+
+        }
+
+        public void InterfaceMessage()
+        {
+            IBus bus = null;
+            #region InterfacePublish
+            var message = new MyEvent { SomeProperty = "Hello world" };
```
Comment (SimonCropp) [inline Snippets/Snippets_3/Routing/BasicOperations.cs]: newing up in an interface snippet?

File: Snippets/Snippets_3/Routing/BasicOperations.cs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    using NServiceBus;
+
+    public class BasicOperations
+    {
+        public void ConcreteMessage()
+        {
+            IBus bus = null;
+            #region InstancePublish
+            var message = new MyEvent { SomeProperty = "Hello world" };
+
```
Comment (SimonCropp) [inline Snippets/Snippets_3/Routing/BasicOperations.cs]: kill the newline

File: Snippets/Snippets_4/Routing/BasicOperations.cs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    using NServiceBus;
+
+    public class BasicOperations
+    {
+        public void ConcreteMessage()
+        {
+            IBus bus = null;
+            #region InstancePublish
+            var message = new MyEvent { SomeProperty = "Hello world" };
+
```
Comment (SimonCropp) [inline Snippets/Snippets_4/Routing/BasicOperations.cs]: kill the newline

File: Snippets/Snippets_4/Routing/BasicOperations.cs [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+            #endregion
+
+        }
+
+        public void InterfaceMessage()
+        {
+            IBus bus = null;
+            #region InterfacePublish
+            var message = new MyEvent { SomeProperty = "Hello world" };
+
```
Comment (SimonCropp) [inline Snippets/Snippets_4/Routing/BasicOperations.cs]: kill the newline

File: Snippets/Snippets_5/Routing/BasicOperations.cs [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
```
Comment (SimonCropp) [inline Snippets/Snippets_5/Routing/BasicOperations.cs]: kill the newline

File: Snippets/Snippets_6/Scanning/ScanningPublicApi.cs [hunk 17]
```diff
@@ -45,5 +45,8 @@ public void Customize(BusConfiguration busConfiguration)
         }
 
         #endregion
+
+        interface IConfigureThisEndpoint
```
Comment (SimonCropp) [inline Snippets/Snippets_6/Scanning/ScanningPublicApi.cs]: what is this for?

File: Snippets/Snippets_6/Snippets_6.csproj [hunk 18]
```diff
@@ -117,15 +124,17 @@
...
   </ItemGroup>
   <ItemGroup>
-    <None Include="app.config" />
     <None Include="Audit\App.config" />
     <None Include="Errors\ErrorQueue\configureErrorQueue.config" />
     <None Include="Errors\FirstLevel\flr.config" />
     <None Include="packages.config">
       <SubType>Designer</SubType>
     </None>
+    <None Include="packages.config" />
```
Comment (SimonCropp) [inline Snippets/Snippets_6/Snippets_6.csproj]: two configs?

File: Snippets/Snippets_6/Scanning/ScanningPublicApi.cs (hunk 17 above)
Comment (andreasohlund) [inline Snippets/Snippets_6/Scanning/ScanningPublicApi.cs]: No idea, removed :)

**Pull request merged**